src/main/java/uk/ac/ic/wlgitbridge/bridge/context/ContextStore.java

```diff
@@ -0,0 +1,132 @@
+package uk.ac.ic.wlgitbridge.bridge.context;
+
+import uk.ac.ic.wlgitbridge.util.Log;
+import java.util.HashMap;
+
+/**
+ * A singleton that manages Project Contexts in the application.
+ * Contexts are wrappers around a ProjectLock, and any other
+ * project-specific resources that need to be managed during
+ * an "operation" on a project.
+ *
+ * Contexts are kept in a ThreadLocal variable, so they can
+ * be re-used once they have been created.
+ *
+ * The intended use is to open a context at the 'top' of a
+ * git operation (push, pull, swap), and lock the project.
+ * Then, deeper in the call-stack the same context can be
+ * retrieved and used.
+ *
+ * (In practice, this enables us to use a Postgres connection
+ * to lock a row in the database, and hold a transaction open,
+ * to be used further down the call stack.)
+ */
+public class ContextStore {
+  private ProjectContextFactory projectContextFactory;
+  static private ContextStore instance;
+  private boolean shuttingDown;
+
+  private ThreadLocal<HashMap<String, ProjectContext>> localContexts;
+
+  private ContextStore(ProjectContextFactory contextFactory) {
+    this.projectContextFactory = contextFactory;
+    this.shuttingDown = false;
+    this.localContexts = new ThreadLocal<HashMap<String, ProjectContext>>();
+  }
+
+  /**
+   * Must be called early in application life-cycle.
+   * - @see GenericProjectContextFactory
+   * - @see PostgresProjectContextFactory
+   */
+  public static void Initialize(ProjectContextFactory contextFactory) throws ContextStoreAlreadyInitialisedException {
+    Log.info("Initialize ContextStore");
+    if (ContextStore.instance != null) {
+        Log.error("ContextStore already initialized");
+      throw new ContextStoreAlreadyInitialisedException();
+    }
+    ContextStore.instance = new ContextStore(contextFactory);
+  }
+
+  /**
+   * Reset the ContextStore, only to be used in tests,
+   * hence the __scary__ underscores
+   */
+  public static void __Reset__() {
+    Log.info("Reset ContextStore");
+    if (ContextStore.instance != null) {
+      ContextStore.Instance().stop();
+      ContextStore.instance = null;
+    }
+  }
+
+  public static ContextStore Instance()  {
+    if (ContextStore.instance == null) {
+      throw new RuntimeException("ContextStore not initialized");
+    }
+    return ContextStore.instance;
+  }
+
+  /**
+   * Cleanup contexts, and pass on the close signal.
+   * Essential for cleanly shutting down the server
+   */
+  public synchronized void stop() {
+    Log.info("Stopping ContextStore");
+    this.clearLocalContexts();
+    if (this.projectContextFactory != null) {
+      this.projectContextFactory.close();
+    }
+  }
+
+  private synchronized void ensureThreadLocalStorage() {
```

Comment:
Just my curious self: is it necessary for methods to be synchronized when they access thread-local storage? One would think that the thread-local variables don't risk being accessed by another thread.

Reply:
Good point, this is probably not required.